modules/local-caching-classloader/src/test/java/org/apache/accumulo/classloader/lcc/LocalCachingContextClassLoaderFactoryTest.java

```diff
@@ -0,0 +1,391 @@
+    assertTrue(fs.mkdirs(new Path("/contextDefs")));
+    final Path invalid = new Path("/contextDefs/InvalidContextDefinitionFile.json");
+    try (FSDataOutputStream out = fs.create(invalid)) {
+      ContextDefinition def = createContextDef("invalid", jarAOrigLocation);
+      out.writeBytes(def.toJson().substring(0, 4));
+    }
+    assertTrue(fs.exists(invalid));
+    final URL invalidDefUrl = new URL(fs.getUri().toString() + invalid.toUri().toString());
+
+    LocalCachingContextClassLoaderFactory factory = new LocalCachingContextClassLoaderFactory();
+    ContextClassLoaderException ex = assertThrows(ContextClassLoaderException.class,
+        () -> factory.getClassLoader(invalidDefUrl.toString()));
+    assertTrue(ex.getMessage().startsWith(
+        "Error getting classloader for context: com.google.gson.stream.MalformedJsonException"));
+  }
+
+  @Test
+  public void testInitial() throws Exception {
+    // Create a new context definition file in HDFS, but with invalid content
+    assertTrue(fs.mkdirs(new Path("/contextDefs")));
+    final Path initial = new Path("/contextDefs/InitialContextDefinitionFile.json");
+    try (FSDataOutputStream out = fs.create(initial)) {
+      ContextDefinition def = createContextDef("initial", jarAOrigLocation);
+      out.writeBytes(def.toJson());
+    }
+    assertTrue(fs.exists(initial));
+    final URL initialDefUrl = new URL(fs.getUri().toString() + initial.toUri().toString());
+
+    LocalCachingContextClassLoaderFactory factory = new LocalCachingContextClassLoaderFactory();
+    ClassLoader cl = factory.getClassLoader(initialDefUrl.toString());
+    @SuppressWarnings("unchecked")
+    Class<? extends test.Test> clazzA =
+        (Class<? extends test.Test>) cl.loadClass("test.TestObjectA");
+    test.Test a1 = clazzA.getDeclaredConstructor().newInstance();
+    assertEquals("Hello from A", a1.hello());
+  }
+
+  @Test
+  public void testUpdate() throws Exception {
+    // Create a new context definition file in HDFS
+    assertTrue(fs.mkdirs(new Path("/contextDefs")));
+    final Path defFilePath = new Path("/contextDefs/UpdateContextDefinitionFile.json");
+    final ContextDefinition def = createContextDef("update", jarAOrigLocation);
+    try (FSDataOutputStream out = fs.create(defFilePath)) {
+      out.writeBytes(def.toJson());
+    }
+    assertTrue(fs.exists(defFilePath));
+    final URL updateDefUrl = new URL(fs.getUri().toString() + defFilePath.toUri().toString());
+
+    final LocalCachingContextClassLoaderFactory factory =
+        new LocalCachingContextClassLoaderFactory();
+    final ClassLoader cl = factory.getClassLoader(updateDefUrl.toString());
+    @SuppressWarnings("unchecked")
+    Class<? extends test.Test> clazzA =
+        (Class<? extends test.Test>) cl.loadClass("test.TestObjectA");
+    test.Test a1 = clazzA.getDeclaredConstructor().newInstance();
+    assertEquals("Hello from A", a1.hello());
+
```

Comment:
prior to the update it would be good to verify D is not around for some unexpected reason

```suggestion
assertThrows(ClassNotFoundException.class, () -> cl.loadClass("test.TestObjectD"));
```

Reply:
Added this in f4387dd